doc: Rework of nrf_security doc

b-gent opened the pull request:
Splitting the one huge RST file into smaller ones.
This should help in maintaining this doc in the future.
Fixing a lot of links that got broken due to structure change.
Also, some small general fixes.

Signed-off-by: Bartosz Gentkowski <bartosz.gentkowski@nordicsemi.no>

NordicBuilder commented:
**Some checks failed. Please fix and resubmit.**

## Nits issues

```
Please remove blank lines at end of 'nrf_security/doc/api.rst'
```



*Tip: The bot edits this comment instead of posting a new one, so you can check the comment's history to see earlier messages.*

b-gent commented:
@frkv  These are all new files so the diff doesn't really help when reviewing this. I suggest building the doc output and going through the pages.

b-gent commented:
https://github.com/nrfconnect/sdk-nrf/pull/3006

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -244,10 +47,11 @@ AES core support can be enabled by setting the :option:`CONFIG_MBEDTLS_AES_C` Kc
 
 .. note::
-   * The `Arm CryptoCell cc310 backend`_ is limited to key sizes of 128 bits.
+   * The :ref:`nrf_security_backends_cc_310` is limited to key sizes of 128 bits.
```
There is no use of the abbreviated form that uses a split betwen cc and 310. 
Please call this cc310 in the link

b-gent replied:
This is now fixed everywhere.

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -261,19 +65,21 @@ AES core support can be enabled by setting setting the :option:`CONFIG_MBEDTLS_A
-   * The `Arm CryptoCell cc310 backend`_ is limited to key sizes of 128 bits.
-   * Enabling the `nrf_oberon backend`_ replaces select internal APIs for AES block encrypt/decrypt and set key operations for encrypt/decrypt.
-   * If both nrf_oberon backend and `Original mbed TLS backend`_ are enabled, the implementation from
+   * The :ref:`nrf_security_backends_cc_310` is limited to key sizes of 128 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -292,19 +98,21 @@ AES cipher modes can be enabled by setting one or more of the following Kconfig
 .. note::
    * AES cipher modes are dependent on enabling AES core support according to `AES configuration`_.
-   * The `Arm CryptoCell cc310 backend`_ is limited to key sizes of 128 bits.
+   * The :ref:`nrf_security_backends_cc_310` is limited to key sizes of 128 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -328,17 +137,19 @@ CMAC can be enabled by setting the :option:`CONFIG_MBEDTLS_CMAC_C` Kconfig varia
 .. note::
-   * The `Arm CryptoCell cc310 backend`_ is limited to key sizes of 128 bits.
-   * If both `nrf_oberon backend`_ and `Original mbed TLS backend`_ is enabled, the implementation from
+   * The :ref:`nrf_security_backends_cc_310` is limited to key sizes of 128 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -357,15 +168,17 @@ AEAD cipher mode support can be enabled by setting one or more of the following
 .. note::
    * AEAD AES cipher modes are dependent on enabling AES core support according to `AES configuration`_.
-   * The `Arm CryptoCell cc310 backend`_ is limited to key sizes of 128 bits.
+   * The :ref:`nrf_security_backends_cc_310` is limited to key sizes of 128 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -381,12 +194,13 @@ AES CCM/CCM* can be enabled by setting the :option:`CONFIG_MBEDTLS_CCM_C` Kconfi
    * AEAD AES cipher modes are dependent on AES core support according to `AES configuration`_.
-   * The `Arm CryptoCell cc310 backend`_ is limited to key sizes of 128 bits.
-   * If both `nrf_oberon backend`_ and `Original mbed TLS backend`_ is enabled, the nrf_oberon backend will
+   * The :ref:`nrf_security_backends_cc_310` is limited to key sizes of 128 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -411,11 +226,12 @@ ChaCha20 support can be enabled by setting the :option:`CONFIG_MBEDTLS_CHACHA20_
-   * For features provided with `Choice`_ support, the enabled backend that is first in order is selected by default.
-   * The ChaCha-Poly implemented by the `Arm CryptoCell cc310 backend`_ does not support incremental operations.
+   * For features provided with :ref:`Choice<nrf_security_backend_config_multiple>` support, the enabled backend that is first in order is selected by default.
+   * The ChaCha-Poly implemented by the :ref:`nrf_security_backends_cc_310` does not support incremental operations.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -427,12 +243,12 @@ Poly1305 can be enabled by setting the :option:`CONFIG_MBEDTLS_POLY1305_C` Kconf
-   * For features provided with `Choice`_ support, the enabled backend that is first in order is selected by default.
-   * The ChaCha-Poly implemented by the `Arm CryptoCell cc310 backend`_ does not support incremental operations.
+   * For features provided with :ref:`Choice<nrf_security_backend_config_multiple>` support, the enabled backend that is first in order is selected by default.
+   * The ChaCha-Poly implemented by the :ref:`nrf_security_backends_cc_310` does not support incremental operations.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -446,15 +262,17 @@ ChaCha-Poly can be enabled by setting the :option:`CONFIG_MBEDTLS_CHACHAPOLY_C`
-   * For features provided with `Choice`_ support, the enabled backend that is first in order is selected by default.
-   * The ChaCha-Poly implementation in `Arm CryptoCell cc310 backend`_ does not support incremental operations.
+   * For features provided with :ref:`Choice<nrf_security_backend_config_multiple>` support, the enabled backend that is first in order is selected by default.
+   * The ChaCha-Poly implementation in :ref:`nrf_security_backends_cc_310` does not support incremental operations.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -464,10 +282,11 @@ DHM can be enabled by setting the :option:`CONFIG_MBEDTLS_DHM_C` Kconfig variabl
 
 .. note::
-   `Arm CryptoCell cc310 backend`_  is limited to key size of <= 2048 bits.
+   :ref:`nrf_security_backends_cc_310`  is limited to key size of <= 2048 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -479,23 +298,26 @@ DHM can be enabled by setting the :option:`CONFIG_MBEDTLS_DHM_C` Kconfig variabl
 
 .. note::
-   * `Arm CryptoCell cc310 backend`_  is limited to key size of <= 2048 bits.
+   * :ref:`nrf_security_backends_cc_310`  is limited to key size of <= 2048 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -619,10 +448,11 @@ RSA support can be enabled by setting the :option:`CONFIG_MBEDTLS_RSA_C` Kconfig
 
 .. note::
-   `Arm CryptoCell cc310 backend`_  is limited to key sizes of <= 2048 bits.
+   :ref:`nrf_security_backends_cc_310`  is limited to key sizes of <= 2048 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backend_config.rst:
```diff
@@ -634,21 +464,21 @@ RSA support can be enabled by setting the :option:`CONFIG_MBEDTLS_RSA_C` Kconfig
-   * For configurations providing `Choice`_ support, the enabled backend that is first in order is selected by default.
-   * `Arm CryptoCell cc310 backend`_  is limited to key sizes of <= 2048 bits.
+   * For configurations providing :ref:`Choice<nrf_security_backend_config_multiple>` support, the enabled backend that is first in order is selected by default.
+   * :ref:`nrf_security_backends_cc_310`  is limited to key sizes of <= 2048 bits.
```
cc_310 split (commented elsewhere)

frkv commented on nrf_security/doc/backends.rst:
```diff
@@ -0,0 +1,89 @@
+   Note that whenever this documentation mentions 'original' mbed TLS, it refers to the open-source `Arm mbed TLS project`_, not the customized version available in Zephyr.
+   KConfig options used to configure this backend include the term ``VANILLA``.
+
+.. _nrf_security_backends_cc_310:
```
cc_310 split (commented elsewhere)

frkv reviewed (approved):
LGTM, but build is failing?
Are you going to try to get this in for 1.4.0?
I didn't see any nrf-sdk PR corresponding to this

I have some additional changes needed for CC312-support (the naming will change from _cc310 to _cc3xx everywhere)

b-gent commented:
> LGTM, but build is failing?
> Are you going to try to get this in for 1.4.0?
> 
> I have some additional changes needed for CC312-support (the naming will change from _cc310 to _cc3xx everywhere) so it would be great to get this in at the earliest

Actually, found two more occurences in glue_layer.rst.

Yes, I want this in for 1.4.0, preferably tomorrow. Red build was simply due to merge conflict, rebased now.

The manifest PR was open but not labeled with the milestone, corrected now.
